mlir/lib/QEC/Transforms/CommuteCliffordTPPR.cpp

```diff
@@ -0,0 +1,407 @@
+        : pauliString(fps.str()), imaginary(fps.imag), negated(fps.value.sign)
+    {
+        // allocated nullptr for correspondingQubits
+        correspondingQubits.resize(pauliString.num_qubits, nullptr);
+    }
+
+    static PauliStringWrapper from_text(std::string_view text)
+    {
+        // check if pauliString contain + or -
+        bool negated = 0;
+        bool imaginary = 0;
+        if (text.starts_with("-")) {
+            negated = true;
+            text = text.substr(1);
+        }
+        else if (text.starts_with("+")) {
+            text = text.substr(1);
+        }
+        if (text.starts_with("i")) {
+            imaginary = true;
+            text = text.substr(1);
+        }
+        auto str = std::string(text);
+        return PauliStringWrapper(std::string(text), imaginary, negated);
+    }
+
+    static PauliStringWrapper from_pauli_words(PauliWords &pauliWords)
+    {
+        std::string pauliStringStr;
+        for (auto pauli : pauliWords) {
+            pauliStringStr += pauli;
+        }
+        return PauliStringWrapper::from_text(pauliStringStr);
+    }
+
+    stim::PauliStringRef<MAX_BITWORD> ref() { return pauliString.ref(); }
+
+    std::string str() const { return pauliString.str(); }
+
+    stim::FlexPauliString flex() { return stim::FlexPauliString(pauliString.ref(), imaginary); }
+
+    PauliWords get_pauli_words()
+    {
+        PauliWords pauliWords;
+        auto str = pauliString.str();
+        for (char c : str) {
+            if (c == 'i' || c == '-' || c == '+')
+                continue;
+            if (c == '_') {
+                pauliWords.push_back("I");
+                continue;
+            }
+            pauliWords.push_back(std::string(1, c));
+        }
+        return pauliWords;
+    }
+
+    bool commutes(PauliStringWrapper &other) { return ref().commutes(other.ref()); }
+};
+
+template <typename T>
+PauliStringWrapper computePauliWords(const llvm::SetVector<Value> &qubits, const T &inOutQubits,
+                                     PPRotationOp &op)
+{
+    // PauliNames initializes with "I" for each qubit
+    PauliWords pauliWords(qubits.size(), "I");
+    std::vector<Value> correspondingQubits;
+    correspondingQubits.resize(qubits.size(), nullptr);
+    for (auto [position, value] : llvm::enumerate(llvm::zip(inOutQubits, op.getPauliProduct()))) {
+        auto qubit = std::get<0>(value);
+        // Find the position of the qubit in array of qubits
+        auto it = std::find(qubits.begin(), qubits.end(), qubit);
+        if (it != qubits.end()) {
+            position = std::distance(qubits.begin(), it);
+            auto pauli = mlir::cast<mlir::StringAttr>(std::get<1>(value)).getValue();
+            pauliWords[position] = pauli;
+            correspondingQubits[position] = qubit;
+        }
+    }
+    auto pauliStringWrapper = PauliStringWrapper::from_pauli_words(pauliWords);
+    pauliStringWrapper.correspondingQubits = correspondingQubits;
```

Comment:
I think this saves us a copy of the vector elements:
```suggestion
    pauliStringWrapper.correspondingQubits = std::move(correspondingQubits);
```

Reply:
That makes sense!